src/stream/mod.rs

```diff
@@ -471,3 +476,231 @@ impl Stream {
+        let reassembly_queue = self.stream.reassembly_queue.lock().await;
+        reassembly_queue.get_num_bytes()
+    }
+}
+
+impl AsyncRead for PollStream<'_> {
+    fn poll_read(
+        mut self: Pin<&mut Self>,
+        cx: &mut Context<'_>,
+        buf: &mut ReadBuf<'_>,
+    ) -> Poll<io::Result<()>> {
+        if buf.remaining() == 0 {
+            return Poll::Ready(Ok(()));
+        }
+
+        let fut = match self.read_fut.as_mut() {
+            Some(fut) => fut,
+            None => {
+                // read into a temporary buffer because `buf` has an unonymous lifetime, which can be
+                // shorter than the lifetime of `read_fut`.
+                let stream = self.stream.clone();
+                let mut temp_buf = vec![0; buf.remaining()];
+                self.read_fut.get_or_insert(Box::pin(async move {
+                    let res = stream.read(temp_buf.as_mut_slice()).await;
+                    match res {
+                        Ok(n) => {
+                            temp_buf.truncate(n);
+                            Ok(temp_buf)
+                        }
+                        Err(e) => Err(e),
+                    }
+                }))
+            }
+        };
+
+        loop {
+            match fut.as_mut().poll(cx) {
+                Poll::Pending => return Poll::Pending,
+                // retry immediately upon empty data or incomplete chunks
+                // since there's no way to setup a waker.
+                Poll::Ready(Err(Error::ErrTryAgain)) => {}
+                // EOF has been reached => don't touch buf and just return Ok
+                Poll::Ready(Err(Error::ErrEof)) => {
+                    self.read_fut = None;
+                    return Poll::Ready(Ok(()));
+                }
+                Poll::Ready(Err(e)) => {
+                    self.read_fut = None;
+                    return Poll::Ready(Err(e.into()));
+                }
+                Poll::Ready(Ok(read_buf)) => {
+                    let len = std::cmp::min(read_buf.len(), buf.remaining());
+                    buf.put_slice(&read_buf[..len]);
+                    self.read_fut = None;
+                    return Poll::Ready(Ok(()));
+                }
+            }
+        }
+    }
+}
+
+impl AsyncWrite for PollStream<'_> {
+    fn poll_write(
+        mut self: Pin<&mut Self>,
+        cx: &mut Context<'_>,
+        buf: &[u8],
+    ) -> Poll<io::Result<usize>> {
+        let (fut, fut_is_new) = match self.write_fut.as_mut() {
+            Some(fut) => (fut, false),
+            None => {
+                let stream = self.stream.clone();
+                let bytes = Bytes::copy_from_slice(buf);
+                (
+                    self.write_fut
+                        .get_or_insert(Box::pin(async move { stream.write(&bytes).await })),
+                    true,
+                )
+            }
+        };
+
+        match fut.as_mut().poll(cx) {
```

Comment:
Note that if `fut_is_new` there's no need to even call `fut.poll()`. `fut.poll()` will be called by the next call to either `poll_write` or `poll_flush`.
This could be used to slightly simplify the implementation.

Reply:
In the current implementation, if there's an error, it will be returned immediately (instead of `Poll::Ready(Ok(buf.len()))`).